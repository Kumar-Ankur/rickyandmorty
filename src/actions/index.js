import * as actionTypes from "../types";

/**
 * @function getCharacters
 * @description action types for fetching characters data from API using saga file and update the state.
 */
export function getCharacters() {
  return { type: actionTypes.GET_CHARACTERS };
}

/**
 * @function getFilteredData
 * @param { Object } data - Filtered data
 * @description - filtered data based on payload value.
 */
export function getFilteredData(data) {
  return { type: actionTypes.GET_FILTERED_DATA, payload: data };
}

/**
 * @function sortAscendingById
 * @description - sort all the character data in ascending order
 */
export function sortAscendingById() {
  return { type: actionTypes.SORT_DATA_BY_ASCENDING };
}

/**
 * @function sortDescendingById
 * @description - sort all the character data in descending order.
 */
export function sortDescendingById() {
  return { type: actionTypes.SORT_DATA_BY_DESCENDING };
}

/**
 * @function addFilteredData
 * @param { String } data
 * @description - add filtered data in string for species category state
 */
export function addFilteredData(data) {
  return { type: actionTypes.ADD_FILTERED_DATA, payload: data };
}

/**
 * @function removeFilteredData
 * @param { String } data
 * @description - remove filtered data in string for species category state
 */
export function removeFilteredData(data) {
  return { type: actionTypes.REMOVE_FILTERED_DATA, payload: data };
}

/**
 * @function addGenderFilteredData
 * @param { String } data
 * @description - add filtered data in string for gender category state
 */
export function addGenderFilteredData(data) {
  return { type: actionTypes.ADD_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function removeGenderFilteredData
 * @param { String } data
 * @description - remove filtered data in string for gender category state
 */
export function removeGenderFilteredData(data) {
  return { type: actionTypes.REMOVE_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function addOriginFilteredData
 * @param { String } data
 * @description - add filtered data in string for origin category state
 */
export function addOriginFilteredData(data) {
  return { type: actionTypes.ADD_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function removeOriginFilteredData
 * @param { String } data
 * @description - remove filtered data in string for origin category state
 */
export function removeOriginFilteredData(data) {
  return { type: actionTypes.REMOVE_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function addSearchText
 * @param { String } data
 * @description - add search data in string for search category state
 */
export function addSearchText(data) {
  return { type: actionTypes.ADD_SEARCH_TEXT, payload: data };
}
