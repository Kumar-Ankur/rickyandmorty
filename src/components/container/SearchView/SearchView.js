import React, { Component } from "react";
import { connect } from "react-redux";
import * as actionCreator from "../../../actions";
import * as constants from "../../../constants";
import "./SearchView.scss";

class SearchView extends Component {

/**
 * @function handleChange
 * @param { Object } event - window event object when user type any word in text box
 * @description - Dispatch action to reducer to update the search text state.
 */
  handleChange = (event) => {
    const { addSearchText } = this.props;
    // dispatch the event to update the search text state.
    addSearchText(event.target.value);
  };
  render() {
    return (
      <div className="form__group field">
        <input
          type="input"
          className="form__field"
          placeholder="Name"
          name="name"
          id="name"
          onChange={(e) => this.handleChange(e)}
          required
        />
        <label htmlFor="name" className="form__label">
          { constants.NAME }
        </label>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addSearchText: (data) => dispatch(actionCreator.addSearchText(data)),
  };
};

export default connect(null, mapDispatchToProps)(SearchView);
