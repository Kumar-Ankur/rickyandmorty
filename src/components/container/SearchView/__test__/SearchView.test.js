import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import SearchView from "../SearchView";
import { mapDispatchToProps } from "../SearchView";
import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "../../../../reducers";
import * as actionCreator from "../../../../actions";

const middleware = createSagaMiddleware();
const mockStore = createStore(rootReducer, applyMiddleware(middleware));

/**
 * Factory function to create a shallowwrapper for the App component.
 * @function setUp
 * @param { Object } props - Component props specific to this setup
 * @param { Object } state - Initial state for setUp
 * @returns { shallowWrapper }
 *
 */
function setUp(props = {}) {
  const wrapper = shallow(<SearchView store={mockStore} {...props} />);
  return wrapper;
}

/**
 * Return shallowWarpper containing node(s) with the given data-test attribute
 * @function findByTestAttr
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper to search within
 * @param {string} val - value of data-test attribute for search
 * @returns { shallowWrapper}
 */
function findByTestAttr(wrapper, val) {
  return wrapper.find(`[data-test="${val}"]`);
}

describe('check search component', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = setUp().dive();
    });

    test('generate snapshot for search component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('check search component render without crash', () => {
      const componentSearch = findByTestAttr(wrapper, "component-search");
      expect(componentSearch.length).toBe(1);
    });

    test('check input field render without error', () => {
      const componentInput = findByTestAttr(wrapper, 'component-input');
      expect(componentInput.length).toBe(1);
    });

    test('check `addSearchText` function call in mapDispatchToProps', () => {
      const dispatchSpy = sinon.spy();
      mockStore.dispatch = dispatchSpy;
      const { addSearchText } = mapDispatchToProps(dispatchSpy);
      addSearchText('Test')
      const lastSpyCall = dispatchSpy.args[0][0];
      const expectedAction = actionCreator.addSearchText('Test');
      expect(lastSpyCall.types).toEqual(expectedAction.types);
    });

    test('check `handleChange` function call on input change', () => {
      const input = findByTestAttr(wrapper, 'component-input');
      const addSearchTextMock = jest.fn();
      const instance = wrapper.instance();
      instance.addSearchText = addSearchTextMock;
      const event = { target: { value: 'test'}}
      input.simulate('change', event);
      instance.addSearchText();
      expect(addSearchTextMock).toHaveBeenCalled();
    })
})