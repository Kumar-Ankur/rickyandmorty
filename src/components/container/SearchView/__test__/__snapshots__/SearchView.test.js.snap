// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check search component generate snapshot for search component 1`] = `ShallowWrapper {}`;
